README
NAME
    Devel::Hints - Access compile-time hints at runtime

VERSION
    This document describes version 0.21 of Devel::Hints, released November
    9, 2010.

SYNOPSIS
        use Devel::Hints ':all';

        LABEL:
        print cop_label();      # 'LABEL'
        # cop_label is only settable on Perl 5.8 or below
        cop_label(0 => 'FOO');  # "goto FOO;" is valid after this point!

        print cop_file();       # same as __FILE__
        print cop_filegv();     # same as \$::{'_<' . __FILE__}
        print cop_stashpv();    # same as __PACKAGE__
        print cop_stash();      # same as \%{__PACKAGE__ . '::'}
        print cop_seq();        # an integer
        print cop_line();       # same as __LINE__

        # cop_arybase() is only settable on Perl 5.8 or below
        print cop_arybase();    # same as $[

        # cop_warnings() is only available to Perl 5.8 or below
        use warnings;
        print cop_warnings();   # same as compile-time ${^WARNING_BITS}

        # cop_io() is only available to Perl 5.7 and Perl 5.8
        use open IO => ':utf8';
        print cop_io();         # same as compile-time ${^OPEN}

        {
            use IO => ':raw';
            print cop_io(1);    # access one uplevel; still ":utf8\0:utf8"
        }

DESCRIPTION
    This module exports the "cop" (code operator) struct as individual
    functions; callers can call them to find out the lexical-scoped hints
    that its block (or statement) is compiled under.

    No functions are exported by default. Each function may take an optional
    positive integer as argument, indicating how many blocks it should walk
    upward to obtain the "cop" members.

    Functions can also take another optional argument, which (if specified)
    *becomes the new value* for the hint, affecting the current statement or
    block's behaviour.

FUNCTIONS
    cop_label
        Label for the current construct.

    cop_file
        File name for the current source file.

    cop_filegv
        Glob reference to the current source filehandle.

    cop_stashpv
        The current package name.

    cop_stash
        Hash reference to the current symbol table.

    cop_seq
        Parse sequencial number.

    cop_arybase
        Array base the calling line was compiled with.

    cop_line
        The line number.

    cop_warnings
        Lexical warnings bitmask, a.k.a. "${^WARNING_BITS}". If no lexical
        warnings are in effect, returns the global warning flags as an
        integer.

    cop_io
        Lexical IO defaults, a.k.a. "${^OPEN}". If no lexical IO layers are
        in effect, an empty string is returned. Always returns "undef" under
        pre-5.7 versions of Perl.

ACKNOWLEDGMENTS
    Thanks to Rafael Garcia-Suarez for demonstrating how to do this with the
    elegant Inline::C code on p5p, which I adapted to this less elegant XS
    implementation.

AUTHORS
    唐鳳 <cpan@audreyt.org>

CC0 1.0 Universal
    To the extent possible under law, 唐鳳 has waived all copyright and
    related or neighboring rights to Devel-Hints.

    This work is published from Taiwan.

    <http://creativecommons.org/publicdomain/zero/1.0>



Changes
* Allow modifying hints of arbitrary coderefs, not just stack frames (doy)

* Fix test failures on perl 5.8 (doy)

[Changes for 0.20 - 2005-12-30]

* LICENSING CHANGE: This compilation and all individual files in it
  are now under the nullary CC0 1.0 Universal terms:

  To the extent possible under law, 唐鳳 has waived all copyright and
  related or neighboring rights to Module-Signature.

* Fix test failures by disabling unsupported operations for Perl 5.10 and 5.12.

* Updated Module::Install::DSL to 1.00.

[Changes for 0.11 - 2005-12-30]

* Fix malformed SIGNATURE file;
  replace ExtUtils::AutoInstall with Module::AutoInstall.
  Reported by: Andreas Koenig
* Update ppport.h to work with latest Perl versions.
* Corrected (harmless) type casting warnings in Hints.xs.

[Changes for 0.10 - 2003-07-27]

* Read/write tests.
* cop_stash can now be assigned \%hash.
* cop_filegv can now be assigned \*glob.
* cop_warnings now correctly returns either a 12-byte string under
  lexical warnings, or a positive integer otherwise.
* cop_io now returns an empty string when there's no value available,
  and undef for perls without perlio.

[Changes for 0.04 - 2003-07-10]

* Now with writable hints!
* Switch to M::I.
* Add ppport.

[Changes for 0.03 - 2003-02-05]

* Now with perl 5.6 and file/stash support.

[Changes for 0.02 - 2003-01-29]

* Add the ability to specify uplevels.

[Changes for 0.01 - 2003-01-29]

* Initial CPAN release.


Hints.pm
package Devel::Hints;
$VERSION = '0.21';

use strict;
use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS);
use 5.006;

require Exporter;
require DynaLoader;

@ISA		= qw( Exporter DynaLoader );
%EXPORT_TAGS	= ( all => \@EXPORT_OK );
@EXPORT_OK	= ( qw(
    cop_label
    cop_stash	cop_stashpv
    cop_file	cop_filegv
    cop_seq
    cop_arybase
    cop_line
    cop_warnings
    cop_io
    ),
);

__PACKAGE__->bootstrap($VERSION);

1;

__END__

=encoding utf8

=head1 NAME

Devel::Hints - Access compile-time hints at runtime

=head1 VERSION

This document describes version 0.21 of Devel::Hints, released
November 9, 2010.

=head1 SYNOPSIS

    use Devel::Hints ':all';

    LABEL:
    print cop_label();	    # 'LABEL'
    # cop_label is only settable on Perl 5.8 or below
    cop_label(0 => 'FOO');  # "goto FOO;" is valid after this point!

    print cop_file();	    # same as __FILE__
    print cop_filegv();	    # same as \$::{'_<' . __FILE__}
    print cop_stashpv();    # same as __PACKAGE__
    print cop_stash();	    # same as \%{__PACKAGE__ . '::'}
    print cop_seq();	    # an integer
    print cop_arybase();    # same as $[
    print cop_line();	    # same as __LINE__

    # cop_warnings() is only available to Perl 5.8 or below
    use warnings;
    print cop_warnings();   # same as compile-time ${^WARNING_BITS}

    # cop_io() is only available to Perl 5.7 and Perl 5.8
    use open IO => ':utf8';
    print cop_io();	    # same as compile-time ${^OPEN}

    {
	use IO => ':raw';
	print cop_io(1);    # access one uplevel; still ":utf8\0:utf8"
    }

=head1 DESCRIPTION

This module exports the C<cop> (code operator) struct as individual
functions; callers can call them to find out the lexical-scoped hints
that its block (or statement) is compiled under.

No functions are exported by default.  Each function may take an
optional positive integer as argument, indicating how many blocks
it should walk upward to obtain the C<cop> members.

Functions can also take another optional argument, which (if specified)
I<becomes the new value> for the hint, affecting the current statement
or block's behaviour.

On perl 5.10 or greater, the first argument to these functions can also be a
coderef. In that case, they return the value for the first statement in the
coderef's body, and if an argument is passed, they set the value for the entire
coderef body. C<cop_line> and C<cop_file> are slightly special here - #line
directives within the coderef will still be respected, and the line will be
offset by the correct amount within the sub.

=head1 FUNCTIONS

=over 4

=item cop_label

Label for the current construct.

=item cop_file

File name for the current source file.

=item cop_filegv

Glob reference to the current source filehandle.

=item cop_stashpv

The current package name.

=item cop_stash

Hash reference to the current symbol table.

=item cop_seq

Parse sequencial number.

=item cop_arybase

Array base the calling line was compiled with.

=item cop_line

The line number.

=item cop_warnings

Lexical warnings bitmask, a.k.a. C<${^WARNING_BITS}>.  If no lexical
warnings are in effect, returns the global warning flags as an integer.

=item cop_io

Lexical IO defaults, a.k.a. C<${^OPEN}>.  If no lexical IO layers
are in effect, an empty string is returned.  Always returns C<undef>
under pre-5.7 versions of Perl.

=back

=head1 ACKNOWLEDGMENTS

Thanks to Rafael Garcia-Suarez for demonstrating how to do this with
the elegant B<Inline::C> code on p5p, which I adapted to this
less elegant XS implementation.

=head1 AUTHORS

唐鳳 E<lt>cpan@audreyt.orgE<gt>

=head1 CC0 1.0 Universal

To the extent possible under law, 唐鳳 has waived all copyright and related
or neighboring rights to Devel-Hints.

This work is published from Taiwan.

L<http://creativecommons.org/publicdomain/zero/1.0>

=cut
